File: receiver_android/app/src/main/java/com/iflytek/lanmediacast/receiver/protocol/RemoteMediaSource.kt
package com.iflytek.lanmediacast.receiver.protocol

import java.net.URI
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.contentOrNull
import kotlinx.serialization.json.jsonObject

data class RemoteMediaTrack(
    val source: JsonObject,
    val httpHeaders: Map<String, String>,
)

data class ValidatedRemoteMediaSource(
    val primaryTrack: RemoteMediaTrack,
    val audioTrack: RemoteMediaTrack?,
)

object RemoteMediaSourceValidator {
    fun validate(source: JsonObject): ValidatedRemoteMediaSource {
        require(source["kind"]?.let(::stringValue) == "url") { "remote media source kind must be url" }
        val name = source["name"]?.let(::stringValue)
        require(name != null && name.isNotBlank() && name.toByteArray(Charsets.UTF_8).size <= 256) {
            "remote media source name is invalid"
        }
        val primaryTrack = validateTrack(source, allowRtsp = true)
        val audioTrack = source["audioTrack"]?.let { element ->
            validateTrack(element.jsonObject, allowRtsp = false, allowAdaptiveManifest = false)
        }
        if (audioTrack != null) {
            require(source["formatHint"] == null) {
                "split remote media tracks must be progressive files"
            }
        }
        return ValidatedRemoteMediaSource(primaryTrack, audioTrack)
    }

    fun validateTrack(
        source: JsonObject,
        allowRtsp: Boolean,
        allowAdaptiveManifest: Boolean = true,
    ): RemoteMediaTrack {
        val rawUrl = source["url"]?.let(::stringValue)
        require(rawUrl != null && rawUrl.toByteArray(Charsets.UTF_8).size <= 4_096) {
            "Remote media URL is invalid"
        }
        val uri = runCatching { URI(rawUrl) }.getOrNull()
        val allowedSchemes = if (allowRtsp) setOf("http", "https", "rtsp") else setOf("http", "https")
        require(uri != null && uri.scheme?.lowercase() in allowedSchemes && !uri.host.isNullOrEmpty()) {
            "Remote media URL is invalid"
        }
        val formatHint = source["formatHint"]?.let(::stringValue)?.lowercase()
        require(formatHint == null || formatHint in setOf("hls", "dash", "rtsp")) {
            "Remote media format is invalid"
        }
        require(allowAdaptiveManifest || formatHint == null) {
            "split remote media tracks must be progressive files"
        }
        require((formatHint == "rtsp") == (uri.scheme.equals("rtsp", ignoreCase = true))) {
            "RTSP format and URL must match"
        }
        val cacheKey = source["cacheKey"]?.let(::stringValue)
        require(cacheKey == null || (cacheKey.isNotEmpty() && cacheKey.toByteArray(Charsets.UTF_8).size <= 256)) {
            "Remote media cache key is invalid"
        }
        val headers = RemoteHttpHeaders.fromSource(source)
        require(!uri.scheme.equals("rtsp", ignoreCase = true) || headers.isEmpty()) {
            "RTSP source cannot contain HTTP headers"
        }
        return RemoteMediaTrack(source, headers)
    }

    private fun stringValue(element: kotlinx.serialization.json.JsonElement): String? {
        val primitive = element as? JsonPrimitive
        return if (primitive?.isString == true) primitive.contentOrNull else null
    }
}

File: sender_flutter/lib/protocol/remote_media_source.dart
import 'dart:convert';

import 'remote_http_headers.dart';

Map<String, Object> validateRemoteMediaSource(Object? rawSource) {
  if (rawSource is! Map) {
    throw const FormatException('remote media source must be an object');
  }
  if (rawSource['kind'] != 'url') {
    throw const FormatException('remote media source kind must be url');
  }
  final Object? rawName = rawSource['name'];
  if (rawName is! String ||
      rawName.trim().isEmpty ||
      utf8.encode(rawName).length > 256) {
    throw const FormatException('remote media source name is invalid');
  }
  final Map<String, Object> normalized = <String, Object>{
    'kind': 'url',
    'name': rawName,
    ...validateRemoteMediaTrack(rawSource, allowRtsp: true),
  };
  final Object? rawAudioTrack = rawSource['audioTrack'];
  if (rawAudioTrack != null) {
    normalized['audioTrack'] = validateRemoteMediaTrack(
      rawAudioTrack,
      allowRtsp: false,
      allowAdaptiveManifest: false,
    );
    if (normalized['formatHint'] != null) {
      throw const FormatException(
        'split remote media tracks must be progressive files',
      );
    }
  }
  return normalized;
}

Map<String, Object> validateRemoteMediaTrack(
  Object? rawTrack, {
  required bool allowRtsp,
  bool allowAdaptiveManifest = true,
}) {
  if (rawTrack is! Map) {
    throw const FormatException('remote media track must be an object');
  }
  final Object? rawUrl = rawTrack['url'];
  if (rawUrl is! String || utf8.encode(rawUrl).length > 4096) {
    throw const FormatException('remote media URL is invalid');
  }
  final Uri? uri = Uri.tryParse(rawUrl);
  final Set<String> allowedSchemes = allowRtsp
      ? const <String>{'http', 'https', 'rtsp'}
      : const <String>{'http', 'https'};
  if (uri == null || !allowedSchemes.contains(uri.scheme) || uri.host.isEmpty) {
    throw const FormatException('remote media URL is invalid');
  }
  final Object? rawFormatHint = rawTrack['formatHint'];
  if (rawFormatHint != null &&
      (rawFormatHint is! String ||
          !const <String>{'hls', 'dash', 'rtsp'}.contains(rawFormatHint))) {
    throw const FormatException('remote media format is invalid');
  }
  if (!allowAdaptiveManifest && rawFormatHint != null) {
    throw const FormatException(
      'split remote media tracks must be progressive files',
    );
  }
  if ((rawFormatHint == 'rtsp') != (uri.scheme == 'rtsp')) {
    throw const FormatException('RTSP format and URL must match');
  }
  final Object? rawCacheKey = rawTrack['cacheKey'];
  if (rawCacheKey != null &&
      (rawCacheKey is! String ||
          rawCacheKey.isEmpty ||
          utf8.encode(rawCacheKey).length > 256)) {
    throw const FormatException('remote media cache key is invalid');
  }
  final Map<String, String> headers = validateRemoteHttpHeaders(
    rawTrack['httpHeaders'],
  );
  if (uri.scheme == 'rtsp' && headers.isNotEmpty) {
    throw const FormatException('RTSP source cannot contain HTTP headers');
  }
  return <String, Object>{
    'url': uri.toString(),
    'formatHint': ?rawFormatHint,
    'cacheKey': ?rawCacheKey,
    if (headers.isNotEmpty) 'httpHeaders': headers,
  };
}
